File: mixin-base/src/main/java/net/yakclient/mixin/base/internal/methodadapter/core/CoreBEPatternMatcher.java
package net.yakclient.mixin.base.internal.methodadapter.core;

import net.yakclient.mixin.base.internal.bytecode.ByteCodeUtils;
import net.yakclient.mixin.base.internal.instruction.Instruction;
import org.objectweb.asm.*;

public class CoreBEPatternMatcher extends CoreMixinPatternMatcher {
    private static final int FOUND_RETURN = 1;
    private static final int FOUND_LAST_RETURN = 2;

    private int returnType;

    public CoreBEPatternMatcher(MethodVisitor visitor, Instruction instructions) {
        super(visitor, instructions);
        this.returnType = Opcodes.RETURN;
    }

    @Override
    public void visitInsn(int opcode) {
        if (ByteCodeUtils.isReturn(opcode)) {
            this.state = FOUND_RETURN;
            this.returnType = opcode;
            return;
        }
        super.visitInsn(opcode);
    }

    @Override
    /* Has cases that will throw */
    public void visitEnd() {
        //Pointless but is nice for cleanup
        if (this.state == FOUND_RETURN)
            this.state = FOUND_LAST_RETURN;
        this.executeInsn();

        super.visitInsn(this.returnType);

        super.visitMaxs(0, 0); /* Should be calculated by the ClassWriter, otherwise it will throw a verify error */
        super.visitEnd();
    }

    @Override
    public void visitInsn() {
        if (state == FOUND_RETURN) {
            state = NOT_MATCHED;
            this.visitInsn(this.returnType);
        }
    }

    @Override
    public void visitLocalVariable(String name, String descriptor, String signature, Label start, Label end, int index) { /* Unused because we have no need to provide debugging tools */ }

    @Override
    public AnnotationVisitor visitLocalVariableAnnotation(int typeRef, TypePath typePath, Label[] start, Label[] end, int[] index, String descriptor, boolean visible) {
        return null; /* We dont need annotations to none-existent references */
    }

    @Override
    public void visitFrame(int type, int numLocal, Object[] local, int numStack, Object[] stack) { /* Should be calculated automatically so again we have no need to provide it */ }

    @Override
    public void visitMaxs(int maxStack, int maxLocals) { /* We dont want Max's to be called before we inject */ }
}

File: base/src/main/kotlin/net/yakclient/mixins/base/internal/methodadapter/core/CoreBEPatternMatcher.kt
package net.yakclient.mixins.base.internal.methodadapter.core

import net.yakclient.mixins.base.internal.bytecode.ByteCodeUtils.isReturn
import net.yakclient.mixins.base.internal.instruction.Instruction
import org.objectweb.asm.*

class CoreBEPatternMatcher(
    visitor: MethodVisitor,
    instructions: Instruction,
) : CoreMixinPatternMatcher(visitor, instructions) {
    private var returnType: Int = Opcodes.RETURN

    override fun visitInsn(opcode: Int) {
        if (isReturn(opcode)) {
            state = FOUND_RETURN
            returnType = opcode
            return
        }
        super.visitInsn(opcode)
    }

    override fun visitEnd() {
        //Pointless but is nice for cleanup
        if (state == FOUND_RETURN) state = FOUND_LAST_RETURN
        this.executeInsn()
        super.visitInsn(returnType)
        super.visitMaxs(0, 0) /* Should be calculated by the ClassWriter, otherwise it will throw a verify error */
        super.visitEnd()
    }

    override fun visitInsn() {
        if (state == FOUND_RETURN) {
            state = NOT_MATCHED
            this.visitInsn(returnType)
        }
    }

    override fun visitLocalVariable(
        name: String,
        descriptor: String,
        signature: String,
        start: Label,
        end: Label,
        index: Int
    ) { }

    override fun visitLocalVariableAnnotation(
        typeRef: Int,
        typePath: TypePath,
        start: Array<Label>,
        end: Array<Label>,
        index: IntArray,
        descriptor: String,
        visible: Boolean
    ): AnnotationVisitor? { return null }

    override fun visitFrame(
        type: Int,
        numLocal: Int,
        local: Array<Any>,
        numStack: Int,
        stack: Array<Any>
    ) { /* Should be calculated automatically so again we have no need to provide it */
    }

    override fun visitMaxs(maxStack: Int, maxLocals: Int) { /* We dont want Max's to be called before we inject */
    }

    companion object {
        private const val FOUND_RETURN = 1
        private const val FOUND_LAST_RETURN = 2
    }
}